Telemetry: Don't attempt to compute route when subresources are used

jakub-pomykala opened the pull request:
~Still a work in progress. Tests are running but failing due to no current solution for getting the correct url when using sub resources.~

Changes in the `TelemetryContainerFilter` class move the logic for getting `route` into a separate method and also add in a check for the list size so that if there are any sub resources being use the `route` will not be modified.

For #25058

jakub-pomykala committed:
Adding test for handling of subresources

Adding in test for route with sub-resources

jakub-pomykala committed:
Adding testRouteWithSubResourceWithQueryParam

Also changed the name for the sub resource as it now also tests for a route with a sub resource and PathParam.

jakub-pomykala committed:
Merge branch 'OpenLiberty:integration' into subresource-tests

Azquelt commented on dev/io.openliberty.microprofile.telemetry.1.0.internal_fat/fat/src/io/openliberty/microprofile/telemetry/internal_fat/apps/jaxrspropagation/route/JaxRsRouteTestServlet.java:
```diff
@@ -111,6 +111,70 @@ public void testRouteWithQueryParam() {
+                        .withAttribute(SemanticAttributes.HTTP_METHOD, "GET")
+                        .withAttribute(SemanticAttributes.HTTP_STATUS_CODE, 200L)
+                        .withAttribute(SemanticAttributes.HTTP_ROUTE, getPath() + "getSubResourceWithPathParam/{id}/details")
+                        .withAttribute(SemanticAttributes.HTTP_TARGET, getPath() + "getSubResourceWithPathParam/{id}/details"));
```
```suggestion
                        .withAttribute(SemanticAttributes.HTTP_TARGET, getPath() + "getSubResourceWithPathParam/myIdForTesting/details"));
```

jakub-pomykala replied:
Done

Azquelt replied:
Looks like this change has got lost somewhere

jakub-pomykala committed:
Changing expected HTTP_URL

Changing from the id param provided to the placeholder that the exptected HTTP_URL should be.

jakub-pomykala committed:
Checking resourceSize in TelemetryContainerFilter

Also refactoring of duplicated code into its own method

jakub-pomykala committed:
Renaming checkForSubResources Method and adding comment

Renamed the method to getRoute as that is the main function of the method also added a comment explaining what the checkResourceSize is doing

jakub-pomykala committed:
Merge remote-tracking branch 'upstream/integration' into subresource-tests

Azquelt commented on dev/io.openliberty.microprofile.telemetry.1.0.internal_fat/fat/src/io/openliberty/microprofile/telemetry/internal_fat/apps/jaxrspropagation/route/JaxRsRouteTestServlet.java:
```diff
@@ -111,6 +111,70 @@ public void testRouteWithQueryParam() {
+        assertThat(serverSpan, isSpan()
+                        .withKind(SpanKind.SERVER)
+                        .withAttribute(SemanticAttributes.HTTP_METHOD, "GET")
+                        .withAttribute(SemanticAttributes.HTTP_STATUS_CODE, 200L)
+                        .withAttribute(SemanticAttributes.HTTP_ROUTE, getPath() + "/getSubResourceWithQueryParam"));
```
Please test `HTTP_TARGET` here as well

Azquelt commented on dev/io.openliberty.microprofile.telemetry.1.1.internal/src/io/openliberty/microprofile/telemetry11/internal/rest/TelemetryContainerFilter.java:
```diff
@@ -145,34 +145,22 @@ public void filter(final ContainerRequestContext request) {
-                    route = template.toTemplate();
-                    ROUTE_CACHE.putRoute(resourceClass, resourceMethod, route);
+                if (route != null) {
+                    currentSpan.setAttribute(SemanticAttributes.HTTP_ROUTE, route);
+                    currentSpan.updateName(request.getMethod() + " " + route);
                 }
 
                 currentSpan.setAttribute(SemanticAttributes.HTTP_ROUTE, route);
-                currentSpan.updateName(request.getMethod()  + " " + route);
+                currentSpan.updateName(request.getMethod() + " " + route);
```
You're setting the attribute and the name both inside and outside of your if block

Azquelt commented on dev/io.openliberty.microprofile.telemetry.1.0.internal/src/io/openliberty/microprofile/telemetry10/internal/rest/TelemetryContainerFilter.java:
```diff
@@ -208,6 +195,36 @@ public void filter(final ContainerRequestContext request, final ContainerRespons
+
+        if (route == null) {
+
+            int checkResourceSize = request.getUriInfo().getMatchedResources().size();
+
+            // Check the resource size using getMatchedResource()
+            // A resource size > 1 indicates that there is a subresource
+            if (checkResourceSize == 1) {
```
```suggestion
            int resourceCount = request.getUriInfo().getMatchedResources().size();

            // Check the resource count using getMatchedResource()
            // A resource count > 1 indicates that there is a subresource
            // We can't currently compute the route correctly when subresources are used
            if (resourceCount == 1) {
```

Please make this change in the 1.1 version of `TelemetryContainerFilter` as well

jakub-pomykala committed:
Removing duplicate setAttribute code from filter

Commit addresses Andrew's comments:
- Adding in extra comment on getRoute method
- Adding in HTTP_TARGET to testRouteWithSubResourceWithQueryParam

Azquelt commented on dev/io.openliberty.microprofile.telemetry.1.0.internal/src/io/openliberty/microprofile/telemetry10/internal/rest/TelemetryContainerFilter.java:
```diff
@@ -153,30 +153,17 @@ public void filter(final ContainerRequestContext request) {
-        } catch (Exception e) {
+        } catch (
+
+        Exception e) {
```
Extra newlines here

jakub-pomykala committed:
Updating HTTP_ROUTE and HTTP_TARGET for testRouteWithSubResourceWithQueryParam

Also updating the HTTP_URL for client side to expect the sub resource

Fixing HTTP_URL in testRouteWithSubResourceWithPathParam

Also removing extra line in TelemetryContainerFilter

jakub-pomykala committed:
Changing testRouteWithSubResourceWithQueryParam and testRouteWithSubResourceWithPathParam

The current behvaiour is different from the ideal behaviour

jakub-pomykala committed:
Commenting out HTTP_TARGET check

Issue #26905 is stopping the correct HTTP_TARGET.

jakub-pomykala commented:
#build
#spawn.fullfat.buckets=io.openliberty.microprofile.telemetry.1.0.internal_fat,io.openliberty.microprofile.telemetry.1.0.internal.jaeger_fat,io.openliberty.microprofile.telemetry.1.0.internal_fat_tck,io.openliberty.microprofile.telemetry.1.1.internal_fat_tck

Azquelt commented:
#libby
